testing/src/test_node.rs

```diff
@@ -0,0 +1,310 @@
+                error!(parent: self.node().span(), "invalid challenge request from {}", peer_addr);
+                return Err(io::ErrorKind::InvalidData.into());
+            };
+
+        // Respond with own challenge request.
+        let own_response = SnarkosMessage::ChallengeResponse(genesis_block_header.clone());
+        trace!(parent: self.node().span(), "sending a challenge response to {}", peer_addr);
+        let msg = own_response.serialize().unwrap();
+        let len = u32::to_le_bytes(msg.len() as u32);
+        conn.writer().write_all(&len).await?;
+        conn.writer().write_all(&msg).await?;
+
+        // Wait for the challenge response to come in.
+        conn.reader().read_exact(&mut buf[..4]).await?;
+        let len = u32::from_le_bytes(buf[..4].try_into().unwrap()) as usize;
+        conn.reader().read_exact(&mut buf[..len]).await?;
+        let peer_response = SnarkosMessage::deserialize(&buf[..len]);
+
+        if let Ok(Message::ChallengeResponse(block_header)) = peer_response {
+            trace!(parent: self.node().span(), "received a challenge response from {}", peer_addr);
+            if block_header.height() == CHALLENGE_HEIGHT && &block_header == genesis_block_header && block_header.is_valid() {
+                // Register the peer's listening address.
+                locked_peers.insert(peer_listening_addr);
+                debug!(parent: self.node().span(), "added {} to my list of peers", peer_listening_addr);
+
+                Ok(conn)
+            } else {
+                error!(parent: self.node().span(), "invalid challenge response from {}", peer_addr);
+                Err(io::ErrorKind::InvalidData.into())
+            }
+        } else {
+            error!(parent: self.node().span(), "invalid challenge response from {}", peer_addr);
+            Err(io::ErrorKind::InvalidData.into())
+        }
+    }
+}
+
+/// Inbound message processing logic for the test nodes.
+#[async_trait::async_trait]
+impl Reading for TestNode {
+    type Message = SnarkosMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        let mut buf = [0u8; 8 * 1024];
+
+        reader.read_exact(&mut buf[..4])?;
+        let len = u32::from_le_bytes(buf[..4].try_into().unwrap()) as usize;
+
+        if reader.read_exact(&mut buf[..len]).is_err() {
+            return Ok(None);
+        }
+
+        match SnarkosMessage::deserialize(&buf[..len]) {
+            Ok(msg) => {
+                info!(parent: self.node().span(), "received a {} from {}", msg.name(), source);
+                Ok(Some(msg))
+            }
+            Err(e) => {
+                error!("a message from {} failed to deserialize: {}", source, e);
+                Err(io::ErrorKind::InvalidData.into())
+            }
+        }
+    }
+
+    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
+        match message {
```

Comment:
Suggestion: move the blocks into functions as this set is likely to grow?

Reply:
makes sense - will add a `TODO`